### ISSUE TITLE
camelcase: new allow option strips underscores before checking name

### ISSUE THREAD

**Issue by lukeapage** [eslint/eslint#11000]
**Tell us about your environment**

* **ESLint Version:** 3.7.0
* **Node Version:** 8
* **npm Version:** 6

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
        "camelcase": [
            "error",
            {
                "properties": "always",
                "allow": ["webpack_public_path", "UNSAFE_componentWillMount", "UNSAFE_componentWillRecieveProps", "UNSAFE_componentWillUpdate"]
            }
        ],
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
__webpack_public_path__ = config.publicPath;
```

**What did you expect to happen?**

No Errors

**What actually happened? Please include the actual, raw output from ESLint.**

An error occurred.

Its quite clear from the code what went wrong:

```
                /*
                 * Leading and trailing underscores are commonly used to flag
                 * private/protected identifiers, strip them
                 */
                const name = node.name.replace(/^_+|_+$/g, ""),
                    effectiveParent = (node.parent.type === "MemberExpression") ? node.parent.parent : node.parent;

                // First, we ignore the node if it match the ignore list
                if (isAllowed(name)) {
                    return;
                }
```

It strips `_` only when there is more than one at the beginning or end. This is quite counter intuitive that I need to also strip them for the allow option.
